typing/uniqueness_analysis.ml

```diff
@@ -264,14 +284,13 @@ end = struct
 
        and as a result, we can delay the projection at `x`.
 
-       The downside of relaxing is the loss of completeness: if we relax too
-       much the program will fail type check. In the extreme case we relax it to
-       Error which fails type check outright (and extremely sound, hehe).
+      On the other hand, the analysis shouldn't infer a [Pexp_ident] to have a
+      too strong usage, as that might be unsound.
```

Comment:
I don't understand this: it should always be sound to assume a use is unique instead of aliased? Or to assume that a use is aliased instead of borrowed? The old text seemed more correct there.

Reply:
For example:
```
let x = .. in
use_uniquely x;
use_uniquely x;
```
The analysis should infer both usages of `x` to be aliased (which later conflicts with `unique_use` on `Texp_ident` as expected). It is unsound to infer the usages to be `unique` (which later agrees with `unique_use` on `Texp_ident`, and doesn't trigger mode error)

Reply:
Okay, I see what you are saying. But I find it confusing: it would not be sound for the analysis to map both uses of `x` to `Aliased` because `Aliased` does not contain a unique barrier. Instead, it would be sound to infer both uses to be unique and then _cast_ them to aliased in a process that forces the unique barrier to be aliased. That is exactly what our `seq` function does.

On the other hand, it seems to me that it would always be sound to infer an aliased usage as unique. This rejects lots of valid programs but I don't see how it could be unsound?

Reply:
When the uniqueness analysis sees a `Texp_ident`, there is a duality that causes all the confusion.

- The `Texp_ident` contains a `unique_use` which contains a `Uniqueness.r` (we will ignore linearity for simplicity). If this mode (we call it `a`) is `Aliased`, it's sound to relax it to `Unique` - we are pretending that the consumer of this `Pexp_ident` is requiring a stricter mode, which is sound.
-  The lexical analysis, not seeing any `unique_use`, might decide that a `Pexp_ident` is the only usage and allow it to be `Unique` (we call it `b`). It's sound to relax it to `Aliased` - we are pretending that there are multiple uses of the value and thus less capability is granted to the usage.

The UA ensures that `b` is stronger than `a`. I.e., the uniqueness capability allowed by the lexical structure, is stronger than the uniqueness capability requested by the consumer of the value.

It's sound to make `b` weaker, or to make `a` stronger. It's the same thing.

I agree the text is currently confusing and I will clarify as above.

Reply:
How does the reasoning in the second bullet work when you relax `Error` to `Aliased` or when you relax `Aliased` to `Unused`?

Reply:
From the perspective of the lexical analysis:

If the analysis decides that a `Texp_ident` is `Aliased`, it's sound to relax that to `Unused`.

To better understand this, imagine we add `Unused` (weaker than `Aliased`) to `Mode.Aliasability`, and we have code:
```
let foo (a @ unused) = () in
foo x
``` 
The lexical analysis will decide the `Texp_ident x` should be `Aliased`. Furthermore, it's sound to relax further to `Unused`, which mean a submode call `x.uniqueness >= Unused`, which succeeds. Of course in general, the relaxing would cause submode failure - but it's always sound to do that.

Reply:
What's the business of `Error`? Here is my attempt. Imagine we add a bottom element to uniqueness and call it `Error`. 
Moreover, our UA is a trivial implementation that behaves like this:
1. its thinks all `Pexp_ident x`, by lexical structure that it doesn't even examine, can enjoy the best uniqueness `Error`. This means we will check `x.uniqueness >= Error` which always succeeds.
2. These `Error` usages combines (via `seq` and `par`, etc.) into a single `Error` in the end, and throws type error

This UA is sound. It's also sound if in step 1 we check `x.uniqueness >= Aliased` (by relaxing Error to Aliased). It would raise even more type error, but it's sound to do so.